### PR TITLE
Automatically generate a public API report for @cadl-lang/compiler

### PR DESCRIPTION
**Pull request by daviwil**
This PR adds an extra step in the `build` script to call [api-extractor](https://api-extractor.com/) to generate a public API report for everything we export from the `@cadl-lang/compiler` library.  This should help with the TypeScript side of our public API review (#233).

Unfortunately `api-extractor` doesn't work well with `export * from "X"` style exports, so to adopt this tool for all our libraries we'll have to go explicitly export all symbols from the public interface.  However, I feel like this actually makes us more aware of what we're exposing, so the added friction is a benefit IMO.

Another benefit of `api-extractor`: it shows us where we're missing TSDoc comments on our public interface so it will be much easier to go sanitize those before release.

If we like how this looks, I can go add this tool to the rest of our libraries across both repos.

### CODE REVIEW

**Comment by nguerrera** [comment]
> However, I feel like this actually makes us more aware of what we're exposing, so the added friction is a benefit IMO.

I agree. It was fun not having to do this while we bootstrapped, but at some point the friction is a necessary thing IMHO.

File: packages/compiler/core/parser.ts [hunk 1]
```diff
@@ -123,7 +123,7 @@ interface UndecoratedListKind extends ListKind {
  * The fixed set of options for each of the kinds of delimited lists in Cadl.
  */
 // eslint-disable-next-line @typescript-eslint/no-namespace
-namespace ListKind {
+export namespace ListKind {
```
Comment (nguerrera) [inline packages/compiler/core/parser.ts:126]: This one surprises me, which public API exposes it? This was intended as a parser implementation detail.

Comment (nguerrera) [reply packages/compiler/core/parser.ts:126]: I can't find any references to ListKind other than the declaration in the .md.

Is it maybe just a weirdness with namespaces? I could refactor it to not be a namespace if so.

**Review by nguerrera — APPROVED** [review]

**Review by timotheeguerin — APPROVED** [review]

**Comment by markcowl** [comment]
@daviwil Can we close this PR?

**Comment by daviwil** [comment]
> @daviwil Can we close this PR?

I forgot about this one, saw it again while scanning over open PRs.  I'll see if I can clean it up and get it merged because I think it's a net positive for the work we're doing.

**Comment by timotheeguerin** [comment]
Going to close this as we have typedoc generated docs for js now.

**Pull request closed without merging**